File: Online Judges/Huxley/576ColorindoEmDuasCores.py
def floodFill(graph, color, now, u):
    color[u] = now
    for v in graph[u]:
        if (color[v] == -1):
            if (not floodFill(graph, color, 1 - now, v)):
                return(0)
        elif (color[v] != 1 - now):
            return(0)
    return(1)

while (True):
    nodes = int(input())
    if (nodes == 0): break
    edges = int(input())
    graph = [[] for i in range(nodes)]
    for i in range(edges):
        u, v = map(int, input().split())
        graph[u] += [v]
        graph[v] += [u]

    color = [-1] * nodes
    if (not floodFill(graph, color, 0, 0)):
        print("NOT BICOLORABLE.")
    else:
        print("BICOLORABLE.")

File: Online Judges/Huxley/576ColorindoEmDuasCores.cpp
#include <bits/stdc++.h>
using namespace std;

int bicolorable(vector<int> graph[], int u, int now, int color[])
{
  if (color[u] != -1) return(color[u] != 1 - now);
  color[u] = now;
  for (auto v: graph[u])
    if (!bicolorable(graph, v, 1 - now, color))
      return(0);
  return(1);
}

int main()
{
  int n, m;
  while (scanf("%d", &n) && n)
  {
    scanf("%d", &m);
    vector<int> graph[n]; int u, v;
    for (int i = 0; i < m; i ++)
    {
      scanf("%d %d", &u, &v);
      graph[u].push_back(v);
      graph[v].push_back(u);
    }

    int color[n]; memset(color, -1, sizeof(color));
    printf("%s.\n", bicolorable(graph, 0, 0, color) ? "BICOLORABLE" : "NOT BICOLORABLE");
  }
  return(0);
}
